daffodil-test/src/test/scala/org/apache/daffodil/layers/CheckDigit.scala

```diff
@@ -0,0 +1,205 @@
+      //
+      // write out the layer data
+      originalOutputStream.write(ba)
+      originalOutputStream.close()
+    }
+  }
+
+  private lazy val suspendableOperation = new SuspendableCheckDigitLayerOperation()
+
+  override def endLayerForUnparse(s: UState): Unit = {
+    suspendableOperation.run(s)
+  }
+
+}
+
+object CheckDigit
+  extends LayerTransformerFactory("checkDigit") {
+
+  override def newInstance(
+    maybeLayerCharsetEv: Maybe[LayerCharsetEv], // if not supplied, taken from the srd.
+    maybeLayerLengthKind: Maybe[LayerLengthKind], // explicit or implicit
+    maybeLayerLengthEv: Maybe[LayerLengthEv], // ignored (with warning ideally)
+    maybeLayerLengthUnits: Maybe[LayerLengthUnits], // ignored (with warning ideally)
+    maybeLayerBoundaryMarkEv: Maybe[LayerBoundaryMarkEv], // ignored (for now, this is a future possibility)
+    srd: SequenceRuntimeData) = {
+
+
+    val layerLengthKind = {
+      if (maybeLayerLengthKind.isEmpty) {
+        srd.SDE("The 'dfdlx:layerLengthKind' property must be defined.")
+      } else {
+        val llk = maybeLayerLengthKind.get
+        llk match {
+          case LayerLengthKind.Explicit => {
+            srd.schemaDefinitionUnless(maybeLayerLengthEv.isDefined, "Property dfdlx:layerLength must be defined.")
+            maybeLayerLengthUnits.toScalaOption match {
+              case Some(LayerLengthUnits.Bytes) => // ok
+              case None => // ok
+              case _ => srd.SDE("Property dfdlx:layerLengthUnits must be 'bytes'.")
+            }
+          }
+          case other => srd.SDE("The dfdlx:layerLengthKind '%s' is not supported.", other)
+        }
+        llk
+      }
+    }
+
+    // To simplify, insist the charset is a constant, 8-bit aligned, SBCS encoding, e.g., ascii or 8859-1.
+
+    val layerCharsetEv: CharsetEvBase =
+      if (maybeLayerCharsetEv.isDefined) maybeLayerCharsetEv.get
+      else srd.encodingInfo.charsetEv
+
+    val bitsCharset = layerCharsetEv.optConstant.getOrElse{
+      srd.SDE("Must have a character set encoding defined.")
+    }
+    val javaCharset: Charset = bitsCharset match {
+      case jbc: BitsCharsetJava if jbc.bitWidthOfACodeUnit == 8 => jbc.javaCharset
+      case _ => srd.SDE(
+        "Charset encoding must be for a byte-aligned single-byte character set, but was encoding '%s', which has width %s bits, and alignment %s bits.",
+        bitsCharset.name, bitsCharset.bitWidthOfACodeUnit, bitsCharset.mandatoryBitAlignment)
+    }
+
+    //
+    // ToDo: We can't issue SDW because we don't have a context object
+    // for that sort of compilation time warning. We *should* have that.
+    // That requires a layer factory method called at schema compile time with the
+    // Sequence term (not just term runtime data) as an argument.
+    //
+    val xformer = new CheckDigitExplicit(srd, javaCharset, maybeLayerLengthEv.get)
+    xformer
+  }
+}
```

Comment:
I'm interesting to see how this all changes with plugability. There's a lot of things being used here that I'd consider internal and wouldn't necessarily want a user to directly access. Not an issue for this PR, though.